Added `OutOfRangeException`

XedinUnknown opened the pull request:
Aims to incorporate solution for #3.

- Added new `OutOfRangeException`.
- Added unit tests for constructors, including testing that they call the parameter-less constructor.
- All exception constructors now normalize the code and message parameters.
- Added new trait for creating invalid argument exceptions in other exceptions - avoiding circular exception dependency.
- 100% test coverage.

XedinUnknown committed:
Now depending on normalization and internationalization helper packages

XedinUnknown committed:
Added tests for `OutOfRangeException`

XedinUnknown committed:
Added `OutOfRangeException`

XedinUnknown committed:
Added PHPStorm project config

XedinUnknown committed:
Removed and tightened some tests

There's no need to test creation of SUT in the functional test,
because the unit test takes care of that; even better, because
in the unit test the original constructor is not fired.

Also, now additionally testing for a Dhii throwable interface.

XedinUnknown committed:
Moved some creation testing into the appropriate test

XedinUnknown committed:
Added test for `CreateInvalidArgumentExceptionCapableTrait`

XedinUnknown committed:
Added `CreateInvalidArgumentExceptionCapableTrait`

XedinUnknown committed:
Now using trait implementation

XedinUnknown committed:
Tightened test

XedinUnknown committed:
Added unit test for constructor

Also shifted creation case into unit test.

XedinUnknown committed:
Now using message and code normalization

XedinUnknown committed:
Removed redundant assertion

XedinUnknown committed:
Tightened param types and default values

Normalization is not possible because exceptions cannot throw
each other to avoid infinite exception loop.

XedinUnknown committed:
Added unit test for constructor

Also, shifted the creation test there

XedinUnknown committed:
Now normalizing some constructor args

XedinUnknown committed:
Now ignoring PHPS workspace config.

XedinUnknown committed:
Corrected return type

mecha commented on src/CreateInvalidArgumentExceptionCapableTrait.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return InvalidArgumentException The new exception.
+     */
+    protected function _createInvalidArgumentException(
```
While I agree with **why** this was introduced in this package, I can't help but feel like _can_ result in re-declaration errors, given that it already exists (same name and all) in another package.

https://github.com/Dhii/exception-helper-base/blob/develop/src/CreateInvalidArgumentExceptionCapableTrait.php

Notice that they are identical. The only differences are the doc block and the alias for `InvalidArgumentException`.

Even without autoloading, this presents some risks. How can we avoid this?

XedinUnknown replied:
Excellent catch. I've renamed it now to avoid the clash.

On a related note, this really brings out the benefits of a possible class indexing tool.

XedinUnknown replied:
By the way, @mecha, the helper factory creates a `Dhii\Exception\InvalidArgumentException`, not a native one, while the trait in this package creates a native PHP exception. That's the difference.

mecha replied:
Yeah 👍 

At least ... now it does 😉

XedinUnknown replied:
Always did. Last commit was made 7 days ago, and was a Composer config change ;)

XedinUnknown committed:
Renamed trait

This is because the helper package already contains a trait with this name.
However, they are not supposed to do the same thing. The one in this
package must create a native PHP exception, while the one in the
helper package must create a Dhii exception.
